### PR TITLE
feat(linking): new linking structure and torrent-backed linking

### LINKED ISSUE #400: Support torrent auto-removal while considering cross-seeds?
There are several applications/scripts that will auto remove torrents based on certain conditions.

For example: https://github.com/jerrymakesjelly/autoremove-torrents

This uses a config format like this to determine deletion candidates

```
qbittorrent:
  client: qbittorrent
  host: <qbit host>
  username: <user>
  password: <pass>
  strategies:
    delete_done_after_30_days:
      categories:
        - done
      remove: seeding_time > 2600000
  delete_data: true
```

Unfortunately none of these applications that I'm aware of take into account cross-seeding. For example I want to be able to do "delete torrents after 28 days of seeding time, as long as all associated cross-seeds have also seeded for 28 days".

I believe that this behaviour would fit well into the cross-seed application by expanding it's ability to deal with cross-seeded data. Since you already have the logic to detect cross seeds, you would just need to run that again on the torrent clients existing torrents (filtered for category/label), apply whatever logic to determine whether the torrents should be deleted, and then call the torrent deletion endpoint.

Just a suggestion, thanks for considering!

I've written a script to do this myself, very hardcoded to qbittorrent though.
```
import requests
import sys

# Script to...
# - find timestamp categories
# - find torrents in timestamp categories
# - set category of these torrents to cross-seed
# - delete timestamp categories
# Shell arguments (mandatory)
# 1. Host of the qbittorrent webui instance, for example http://localhost:8098
# 2. webui username
# 3. webui password
# This is only necessary because SeedCross inexplicably adds cross-seeding torrents
#    using a timestamp category
# Run via cronjob

# Check for correct argument count
if len(sys.argv) < 4:
    print('Provide qbittorent host, username, and password arguments')
    sys.exit()

qbitHost = sys.argv[1]
user = sys.argv[2]
password = sys.argv[3]

# Create request session that shares cookies
session = requests.Session()

# Login
loginUrl = qbitHost + '/api/v2/auth/login'
loginPayload = { 'username': user, 'password': password }
headers = { 'Referer': qbitHost }
login = session.post(loginUrl, data = loginPayload, headers = headers)

if not login.ok:
    print("Failed to login")
    sys.exit()


# Get torrents
torrentListRes = session.get(qbitHost + '/api/v2/torrents/info')
if not torrentListRes.ok:
    print("Failed to get torrent list")
    sys.exit()
torrents = torrentListRes.json()

# Group torrents by content_path
# Matching content_path means they are cross-seeding
torrentsByPath = {}
for torrent in torrents:
    contentPath = torrent['content_path']
    if contentPath not in torrentsByPath:
        torrentsByPath[contentPath] = []
    torrentsByPath[contentPath].append(torrent)

seedLimitSeconds = 24 * 60 * 60 * 28 # 28 days
seedHashesToDelete = set()
deleteCategories = ['done']
for torrentsWithMatchingPath in torrentsByPath.values():
    allInDeleteCategories = all(map(
        lambda torrent: torrent['category'] in deleteCategories,
        torrentsWithMatchingPath
    ))
    allSeededForSeedLimit = all(map(
        lambda torrent: torrent['seeding_time'] > seedLimitSeconds,
        torrentsWithMatchingPath
    ))
    if allInDeleteCategories and allSeededForSeedLimit:
        seedHashesToDelete.update(map(
            lambda torrent: torrent['hash'],
            torrentsWithMatchingPath
        ))

# Send delete request
if len(seedHashesToDelete) > 0:
    deleteTorrentsPayload = { 'hashes': '|'.join(seedHashesToDelete),
                            'deleteFiles': 'true' }
    deleteRes = session.post(qbitHost + '/api/v2/torrents/delete', data=deleteTorrentsPayload)
```

### PR DESCRIPTION
**Pull request by zakkarry**
adds functionality to link data-based searches to tracker-specific subdirectories inside of linkDir

fixes nested data not linking to single file torrents

adds torrent backed linking support

potentially close #400

### CODE REVIEW

**Comment by zakkarry** [comment]
> I dont know if I like it, wont it change from the category destination (in case of qbit)?

switch to deluge :P

File: src/action.ts [hunk 1]
```diff
@@ -41,22 +41,30 @@ export async function performAction(
...
-
+	let trackerLinkDir: string;
+	if (linkDir) {
+		trackerLinkDir = join(linkDir, tracker);
+	}
+	if (!existsSync(trackerLinkDir)) {
+		mkdirSync(trackerLinkDir, {
+			recursive: true,
+		});
+	}
```
Comment (mmgoodnow) [inline src/action.ts:52]:
```suggestion
		mkdirSync(trackerLinkDir, {
			recursive: true,
		});
```

File: src/action.ts [hunk 2]
```diff
@@ -41,22 +41,30 @@ export async function performAction(
 	tracker: string
 ): Promise<ActionResult> {
 	const { action, linkDir } = getRuntimeConfig();
-
+	let trackerLinkDir: string;
+	if (linkDir) {
+		trackerLinkDir = join(linkDir, tracker);
+	}
```
Comment (mmgoodnow) [inline src/action.ts:47]: small nit, using a ternary here allows this to be marked const:

```suggestion
	const trackerLinkDir = linkDir ? join(linkDir, tracker) : undefined;
	}
```

**Review by mmgoodnow — COMMENTED** [review]
one thing I'm thinking about is - where do these tracker "names" come from? can we be guaranteed they're stable/consistent? probably…

what would you think of using infohash? would that be too ugly? it feels nice to give each torrent its own "sandbox"

**Comment by zakkarry** [comment]
> what would you think of using infohash? would that look too ugly in the clients? it feels nice to give each torrent its own "sandbox"

I believe tracker name comes from prowlarr/jackett, so it would be whatever the user has it named in their aggregator - it's returned in the XML and parsed in parseXMLResults or whatever the function is named. InfoHash is rather ugly and lengthy relative. I mean, most people probably don't go digging through their files like that, but I would like to keep it somewhat aesthetically pleasing and manageable.

**Review by mmgoodnow — DISMISSED** [review]

**Comment by zakkarry** [comment]
Potentially pending major version bump.

**Comment by mmgoodnow** [comment]
@zakkarry would you mind putting the torrent-backed linking stuff in a separate PR?

**Comment by zakkarry** [comment]
I'll do it later today, about to head out.

**Comment by zakkarry** [comment]
Should I PR to this PR? or what, because otherwise it seems like things could get pretty complicated rebasing if one is merged before the other.

Not sure how to do this, and I sent you a DM about the possibility of changing deluge.ts and the deluge refactor being merged prior to the torrent backed linking, and the mess that would be.

Off to nephew's party though, so message me if you get on.

**Comment by zakkarry** [comment]
Got this branch back to data backed only, and made another branch for torrent linking based on this one.

Will sort out the PR for it later.

**Review by mmgoodnow — DISMISSED** [review]
I wouldn't really consider this a breaking change, but should we merge it into v6 instead of master just in case?

also, one thing worth mentioning - how is this gonna work for people that already have existing links?

**Review by mmgoodnow — APPROVED** [review]

**Pull request merged**